src/driver-marsgaming/marsgaming-buttons.c

```diff
@@ -0,0 +1,425 @@
+#include "marsgaming-buttons.h"
+
+struct marsgaming_button_action_mapping {
+	enum marsgaming_button_action marsgaming_action_id;
+	struct ratbag_button_action ratbag_action;
+};
+
+/**
+ * This array contains only the actions that can be statically generated.
+ * This array is used in the marsgaming_button_action_lookup function.
+ * For any action that cannot be statically generated (has variables in it)
+ * we use the rest of the marsgaming_button_action_* functions
+ */
+static const struct marsgaming_button_action_mapping marsgaming_mm4_button_action_mapping[] = {
+	{ MARSGAMING_MM4_ACTION_LEFT_CLICK, BUTTON_ACTION_BUTTON(1) },
+	{ MARSGAMING_MM4_ACTION_RIGHT_CLICK, BUTTON_ACTION_BUTTON(2) },
+	{ MARSGAMING_MM4_ACTION_MIDDLE_CLICK, BUTTON_ACTION_BUTTON(3) },
+	{ MARSGAMING_MM4_ACTION_BACKWARD, BUTTON_ACTION_BUTTON(4) },
+	{ MARSGAMING_MM4_ACTION_FORWARD, BUTTON_ACTION_BUTTON(5) },
+	{ MARSGAMING_MM4_ACTION_DPI_SWITCH, BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_RESOLUTION_CYCLE_UP) },
+	{ MARSGAMING_MM4_ACTION_DPI_MINUS, BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_RESOLUTION_DOWN) },
+	{ MARSGAMING_MM4_ACTION_DPI_PLUS, BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_RESOLUTION_UP) },
+	{ MARSGAMING_MM4_ACTION_PROFILE_SWITCH, BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_PROFILE_CYCLE_UP) }
+};
+
+static struct ratbag_button_action
+marsgaming_button_action_lookup(struct ratbag_button *button,
+				const struct marsgaming_button_info *button_info)
+{
+	const struct marsgaming_button_action_mapping *button_mapping;
+	ARRAY_FOR_EACH(marsgaming_mm4_button_action_mapping, button_mapping) {
+		if (button_mapping->marsgaming_action_id == button_info->action)
+			return button_mapping->ratbag_action;
+	}
+	return (struct ratbag_button_action)BUTTON_ACTION_UNKNOWN;
+}
+
+static int
+ratbag_button_macro_from_marsgaming_combo_keycode(struct ratbag_button *button, unsigned int key1, unsigned int key2, unsigned int modifiers);
```

Comment:
These function names are misleading as they are provided by the marsgaming driver and not libratbag.
This one could be `marsgaming_ratbag_button_macro_from_combo_keycode` for example.

Reply:
I renamed all functions that started with `ratbag_` to `marsgaming_` to avoid confusion.